### DISCUSSION TITLE
Pending state not returned when user requests to join team

### DISCUSSION THREAD

**Discussion by AlfieBrooks** [community/community#24369] [Apps, API and Webhooks]
<p>Hi, I’m using Octokit to request the current membership status of a user in an organisation team using</p>
<pre><code class="lang-auto">const response = await octokit.teams.getMembershipForUserInOrg({
      org,
      team_slug,
      username,
    });
</code></pre>
<p>Tried using these endpoints:</p>
<ul>
<li><a href="https://docs.github.com/en/rest/reference/teams#get-team-membership-for-a-user" rel="nofollow noopener">https://docs.github.com/en/rest/reference/teams#get-team-membership-for-a-user</a></li>
<li><a href="https://docs.github.com/en/rest/reference/teams#list-pending-team-invitations" rel="nofollow noopener">https://docs.github.com/en/rest/reference/teams#list-pending-team-invitations</a></li>
</ul>
<p>When that user is in the team it returns a 200 with the state as <code>active</code> and returns a 404 when the user isn’t in the team. However, according to the docs, it should return a 200 when they are in the pending state too, When that user clicks the “Request to join team” button on GitHub it doesn’t seem to set them into a <code>pending</code> state and still returns 404.</p>
<p>Does the pending state only apply for invites? and doesn’t work when requesting to join?</p>

**Comment by nethgato** [comment]
<p>Hi, hi <a class="mention" href="/alfiebrooks">@AlfieBrooks</a> :wave:</p>
<p>I went ahead and reproduced this by:</p>
<ul>
<li>Having an Org user request to join my Team</li>
<li>Inviting a user <em>not in the Org</em> to join my Team (users in an Org simply get added to a Team, and there is no invitation/pending status)</li>
</ul>
<p>Between each step, I tried each endpoint:</p>
<ul>
<li>Get team membership for a user</li>
<li>List pending team invites</li>
</ul>
<hr>
<p>What I found, is that <strong>when the user requested access to the Team</strong>…</p>
<p><code>/orgs/{org}/teams/{team_slug}/memberships/{username}</code> returns a 404:</p>
<pre><code class="lang-auto">{
    "message": "Not Found",
    "documentation_url": "https://docs.github.com/rest/reference/teams#get-team-membership-for-a-user"
}
</code></pre>
<p>…and <code>/orgs/{org}/teams/{team_slug}/invitations</code> endpoint returns:</p>
<p>…an empty array:</p>
<pre><code class="lang-auto">➜  ~ curl -L -X GET 'https://api.github.com/orgs/{org}/teams/{team_name}/invitations' \
-H 'Authorization: bearer KEY' \
-H 'Accept: application/vnd.github.v3+json'
[
]
</code></pre>
<hr>
<p><strong>When the invitation was sent to a user not already within the Org to join the Team</strong>…</p>
<p><code>/orgs/{org}/teams/{team_slug}/memberships/{username}</code> returns:</p>
<pre><code class="lang-auto">{
    "state": "pending",
    "role": "member",
    "url": "https://api.github.com/organizations/{org_id}/team/{team_id}/memberships/{username}"
}
</code></pre>
<p><code>/orgs/{org}/teams/{team_slug}/invitations</code> returns:</p>
<pre><code class="lang-auto">[
    {
        "id": {id},
        "node_id": "{node_ID}",
        "login": "username",
        "email": null,
        "role": "direct_member",
        "created_at": "2020-09-28T12:04:22.000-07:00",
        "inviter": {
            "login": "nethgato",
            "id": {id},
            "node_id": "{node_id}",
            "avatar_url": "{avatar_url}",
            "gravatar_id": "",
            "url": "https://api.github.com/users/nethgato",
            "html_url": "https://github.com/nethgato",
            "followers_url": "https://api.github.com/users/nethgato/followers",
            "following_url": "https://api.github.com/users/nethgato/following{/other_user}",
            "gists_url": "https://api.github.com/users/nethgato/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/nethgato/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/nethgato/subscriptions",
            "organizations_url": "https://api.github.com/users/nethgato/orgs",
            "repos_url": "https://api.github.com/users/nethgato/repos",
            "events_url": "https://api.github.com/users/nethgato/events{/privacy}",
            "received_events_url": "https://api.github.com/users/nethgato/received_events",
            "type": "User",
            "site_admin": true
        },
        "team_count": 1,
        "invitation_teams_url": "https://api.github.com/organizations/{org_id}/invitations/{invite_id}/teams"
    }
]
</code></pre>
<hr>
<p>So to your question:</p>
<blockquote>
<p>Does the pending state only apply for invites? and doesn’t work when requesting to join?</p>
</blockquote>
<p>That really does seem to be what’s happening, but it’s not clear that this is expected behavior. Now that I’ve documented these steps here, I’m going to followup with our team to find out if we’re bumping up against unexpected behavior, or perhaps there was something done incorrectly in my reproduction steps.</p>
<p>Thank you for your patience in the meantime! :bow:</p>

**Comment by nethgato** [comment, marked as answer]
<p>Hi again <a class="mention" href="/alfiebrooks">@AlfieBrooks</a>  o/</p>
<p>How are you going? I just wanted to followup here since I did get in touch with our teams responsible for maintenance of those endpoints and do have some clarification to add to my post above.</p>
<p>When a user has a  <em>pending invitation</em>  to join a team, we consider their membership to be pending. When a user has  <em>requested to join</em>  a team, we consider them to have an inactive membership, so that’s why we see the 404 and empty array response.</p>
<p>We don’t necessarily want to treat a membership request to be a pending membership, as users could essentially request to join any number of teams without any guarantee that their requests would be accepted.</p>
<p>So from that, I think there may be some pending cleanup that we could do, here:</p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/free-pro-team@latest/rest/reference/teams#get-team-membership-for-a-user" target="_blank" rel="noopener">docs.github.com</a>
  </header>
  <article class="onebox-body">
    <img src="" class="thumbnail" width="16" height="16">

<h3><a href="https://docs.github.com/en/free-pro-team@latest/rest/reference/teams#get-team-membership-for-a-user" target="_blank" rel="noopener">Teams - GitHub Docs</a></h3>



  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>…where:</p>
<blockquote>
<p>Response if user has a pending membership with team.</p>
</blockquote>
<p>…could be a bit ambiguous.</p>
<p>Curious to know what you think, but also want to thank you for elevating this and hope that the verification of expected behavior is valuable to you!</p>
<p>Let us know. :bow:</p>

**Comment by AlfieBrooks** [comment]
<p>Hi <a class="mention" href="/nethgato">@nethgato</a>,</p>
<p>Thanks for the response! Yeah, your findings are exactly what I was doing to reproduce this issue. It wasn’t clear for me originally that the pending state was only when a user is invited to a team and not for when requesting to join a team.</p>
<p>That’s understandable that you wouldn’t want to return a pending state when they request if there are thousands of requests to a team. Thank you for confirming that for me!</p>
<p>As you also mentioned, it would be good to get the docs updated to explain what counts as a pending state for a user in a team (Only having been invited to a team).</p>
<p>Thanks again,<br>
Alfie</p>

**Comment by nethgato** [comment]
<p>Hey <a class="mention" href="/alfiebrooks">@AlfieBrooks</a> cheers!</p>
<p>So for the docs update, there’s a couple things/options. One, I can simply open an internal issue with our docs team to suggest an update. Two, a submission from you (or anyone else reading with similar requests) via our feedback form:</p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://support-assets.github.com/assets/favicon-3dcd2c977a91b2e3f396b300c5ff0d572295ca632ddb7ba1adcacc1ea68dd5c0.ico" class="site-icon" width="16" height="16">
      <a href="https://support.github.com/contact/feedback" target="_blank" rel="noopener">support.github.com</a>
  </header>
  <article class="onebox-body">
    <img src="" class="thumbnail" width="16" height="16">

<h3><a href="https://support.github.com/contact/feedback" target="_blank" rel="noopener">Share feedback - GitHub Support</a></h3>



  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>I’m happy for us to do either, or both! If you feel there is unclear language, both is probably best. If you have a preference for me submitting that directly, happy to do that, too!</p>
<p>LMK what you think :bow:</p>
